### DISCUSSION TITLE
How would you filter out duplicate features when using an MVTLayer

### DISCUSSION THREAD

**Discussion by DrewVartanian** [visgl/deck.gl#7446] [Q&A]
I'm using an MVTLayer to show zip codes on a map, and since I'm showing the borders around the zip codes, I needed to create the layers with no clipping set (otherwise the tile borders get drawn as well).  This means that a zip code can be on multiple tiles, since it will include the entire zip code without clipping, and it gets rendered onto the map multiple times.  The overlapping polygons are an issue both because it seems to affect performance, and they end up making the fill color effectively less opaque, as I have an opacity set on the fill color.

I'm wondering if there is a way to filter out these duplicate zip codes coming from different tilesets on load.  I've tried playing around with DataFilterExtension, the dataTransform option, and loadOptions.transforms, but couldn't get anything to work.  Any advice or suggestions are appreciated!

**Comment by Pessimistress** [comment]
> since I'm showing the borders around the zip codes, I needed to create the layers with no clipping set (otherwise the tile borders get drawn as well)

Can you elaborate on this? The MVTlayer is designed to handle features split across multiple tiles.

**Reply by Pessimistress** [reply]
Might be what you need: https://github.com/visgl/deck.gl/issues/5192#issuecomment-734876226

**Reply by DrewVartanian** [reply]
So my original issue was similar to what was seen in [#5192](https://github.com/visgl/deck.gl/issues/5192#issuecomment-734876226).  Changing the buffer didn't help in my case, but I found the CLIP option seen [here](https://gdal.org/drivers/vector/mvt.html) under the Opening Options section, which seemed to help.  My loose understanding of what it does is if a polygon spreads across multiple tiles, it will include the whole shape in both tiles, rather than just the part of the polygon that is in any given tile.  This allows the [MVTLayer](https://deck.gl/docs/api-reference/geo-layers/mvt-layer) to draw the shape as a whole, rather than in parts, avoiding an intersection border where the shape crosses between two tiles.  Unfortunately, this means if both tiles are loaded, the polygon will be drawn twice, thus me looking to filter out the duplicate polygons, each coming from a different tile.

My use of the MVTLayer looks something like this:
```
let overlay = new GoogleMapsOverlay({
  layers: [
    new MVTLayer({
      id: 'MVTLayer',
      minZoom: 0,
      maxZoom: 15,
      data: url,
      lineWidthUnits: 'pixels',
      getLineWidth: f => 1,
      getLineColor: f => getColorByFeature(f, 25),
      getFillColor: f => [...getColorByFeature(f, 25), 70],
    })
  ]
});

overlay.setMap(map);
```

**Reply by kylebarron** [reply]
The MVT Layer is designed to clip for you. It essentially uses the [clip extension](https://deck.gl/docs/api-reference/extensions/clip-extension) under the hood iirc. You could either clip the tiles to tile extent without a buffer when you generate them via gdal, or you can clip them to tile extent when rendering. You should not need to de-duplicate actual features

**Reply by DrewVartanian** [reply]
Thanks for the reply!  I'm new to all of this, so if anything I ask is obvious or misdirected, my apologies.  I'm currently using [tippecanoe](https://github.com/mapbox/tippecanoe) and [ogr2ogr](https://gdal.org/programs/ogr2ogr.html) to generate the pbf files/folder structure from geojson, and [MVTLayer](https://deck.gl/docs/api-reference/geo-layers/mvt-layer) to display it on Google Maps.  Do you know how I would either clip the tiles to tile extend without a buffer or how to clip them to tile extend when rendering?

**Reply by Pessimistress** [reply]
The layer does it by default. What version is your deck.gl?

**Reply by DrewVartanian** [reply]
I'm using deck.gl 8.8.20.  To generate the pbf files, I run
```
tippecanoe -o byZip.mbtiles -l mylayer --read-parallel -pk -pc -Z 6 -z 10 byZip.geojson
```
followed by
```
ogr2ogr -f MVT data byZip.mbtiles -dsco MINZOOM=6 -dsco MAXZOOM=10 -oo CLIP=NO
```

When I ran this without the -pc in tippecanoe and without the -oo CLIP=NO, the results looked like the attached images.  After adding the No Clipping settings to tippecanoe and ogr2ogr, the grid you can see in the images when away, but that's how I ended up with the duplicate features.

<img width="1056" alt="Screenshot 2022-11-22 at 6 20 25 PM" src="https://user-images.githubusercontent.com/11863081/203440904-06cbcfea-b069-4e3d-a027-91975f5a27a3.png">
<img width="1056" alt="Screenshot 2022-11-22 at 6 20 15 PM" src="https://user-images.githubusercontent.com/11863081/203440923-b51925d3-20ba-4b37-b507-f0f74def2453.png">

**Reply by Pessimistress** [reply]
@kylebarron may provide more insight about tippecanoe, but I suspect your tiles are not generated correctly.

Here is a demo showing MVTLayer with the automatic clipping disabled. You can see how the buffer extends beyond the bounds of each tile. https://codepen.io/Pessimistress/pen/vYrdojX?editors=0010

**Reply by kylebarron** [reply]
I also suspect your tiles are not generated correctly. I don't know what your second `ogr2ogr` step is doing and I don't think it's necessary

You can create a directory instead of an mbtiles container with the `-e directory_name` flag instead of `-o name.mbtiles`.
If you want to prototype locally from a directory of mvt files, use 
```
tippecanoe -e directory_name --no-tile-compression
```
with something like `python -m http.server` on that directory

If you want to use this data on a remote server in production, leave off the `--no-tile-compression` and set the correct content encoding headers on the bucket. (Otherwise you'll hit tons of [`unimplemented type 3` errors](https://github.com/mapbox/mapbox-gl-js/issues/1567#issuecomment-144783814))

**Reply by DrewVartanian** [reply]
@kylebarron That worked perfectly!  Not only did it fix the issue, but I'm able to remove an otherwise unused tool (ogr2ogr) from my code.  Thanks to @Pessimistress too for taking a look too, I appreciate all the help!
